File: src/wp/MBTilesActionsFileImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using System.IO.IsolatedStorage;
using WPCordovaClassLib.Cordova.JSON;
using MBTilesPlugin;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

/// <summary>
/// the namespace of MBTilesPlugin </summary>
namespace MBTilesPlugin
{
    /// <summary>
    /// class MBTilesActionsFileImpl</summary>
    class MBTilesActionsFileImpl : MBTilesActionsGenImpl
    {
        
        private string mapDirectory = null;

        public MBTilesActionsFileImpl()
            : base()
        {
            directory = Path.Combine(Windows.Storage.ApplicationData.Current.LocalFolder.Path, "maps//");
        }

        public override void open(string name)
	    {
            close();
            try
            {
                using (IsolatedStorageFile isoFile = IsolatedStorageFile.GetUserStoreForApplication())
                {
			        /// test if directory exist
                    string path = getDirectory(); 
                    if (isoFile.DirectoryExists(path))
                    {
                        mapDirectory = path;
                    }
                }
            } 
            catch(Exception) {
                // Add log
            }
        }

        public override bool isOpen()
	    {
		    return (this.mapDirectory != null);
	    }

	    public override void close()
	    {
		  this.mapDirectory = null;
	    }

        /// <summary>
        /// get metadata in JObject format</summary>
  	/// <returns>
	/// the JObject repsenting the metadata</returns>
        private JObject getMetadataObject()
        {
            JObject obj = new JObject(); 
            string filePath = mapDirectory + "\\metadata.json";
            string text = readText(filePath);
	        if (text != null) {
		       
		        metadata[] list = JsonHelper.Deserialize<metadata[]>(text);
			    
			    foreach (metadata data in list)
			    {
				    obj.Add(data.name, data.value);
			    }
		    }
            return obj;
        }

        public override string getMetadata()
	    {
            JObject obj = getMetadataObject();
            return obj.ToString(); ;
		}

        public override minzoom_output getMinZoom()
	    {
            minzoom_output result = null;
            try
            {
                List<int> zoomLevels = getZoomLevels();

                if (zoomLevels.Count > 0)
                {
                    result = new minzoom_output(zoomLevels.ElementAt(0));
                }
            }
            catch (Exception)
            {
                // Add log
            }

            return result;
		}

        public override maxzoom_output getMaxZoom()
	    {
            maxzoom_output result = null;
            try
            {
                List<int> zoomLevels = getZoomLevels();

                if (zoomLevels.Count > 0)
                {
                    result = new maxzoom_output(zoomLevels.ElementAt(zoomLevels.Count - 1));
                }
            }
            catch (Exception)
            {
                // Add log
            }
            return result;
		}

        public override tiles_output getTile(int zoomLevel, int column, int row)
	    {
            tiles_output result = null;
            int currentZoomLevel = zoomLevel;
            try
		    {
			    maxzoom_output max = getMaxZoom();
                if (max != null)
                {
                    int maxZoom = max.value;

                    if (zoomLevel > maxZoom)
                    {
                        currentZoomLevel = maxZoom;
                    }
                }
                
		    }
		    catch (Exception)
		    {
			}

            try
		    {
                JObject metadata = getMetadataObject();
                if (metadata != null)
                {
                    string name = (string)metadata.GetValue(ConstantMbTilePlugin.KEY_NAME);
                    string version = (string)metadata.GetValue(ConstantMbTilePlugin.KEY_VERSION);
                    string format = (string)metadata.GetValue(ConstantMbTilePlugin.KEY_FORMAT);
		            if (name != null && version != null && format != null)
                    {
			            /// get path of tiles
			            string tileFile = this.mapDirectory + "\\" + version + "\\" + name + "\\" +
			                                    currentZoomLevel.ToString() + "\\" + column.ToString() + "\\" + row.ToString() + "." + format;
			            /// encode raw
			            string raw = readText(tileFile);
			            if (raw != null) {
                            string data = ConstantMbTilePlugin.Base64Encode(raw);
				            result = new tiles_output(data);
			            }
		            }
                }
            }
            catch(Exception)
            {

            }

            return result;

		}


        /// <summary>
        /// read raw of given file</summary>
  	/// <returns>
	/// return the raw of file in string</returns>
	/// <param name="filePath"> the path of file</param>
        private string readText(string filePath) 
        {
            string text = null;
            try
            {
                using (IsolatedStorageFile isoFile = IsolatedStorageFile.GetUserStoreForApplication())
                {
                    if (isoFile.FileExists(filePath))
                    {
                        // read text
                        using (IsolatedStorageFileStream reader = isoFile.OpenFile(filePath, FileMode.Open, FileAccess.Read))
                        {
                            var buffer = new byte[(int)reader.Length];

                            reader.Seek(0, SeekOrigin.Begin);
                            reader.Read(buffer, 0, buffer.Length);

                            Encoding encoding = Encoding.GetEncoding("UTF-8");
                            text = encoding.GetString(buffer, 0, buffer.Length);
                        }
                    }
                }
            } 
            catch(Exception)
            {

            }
            return text;
        }

        /// <summary>
        /// get all zoom level</summary>
  	    /// <returns>
	    /// return a list of zoom level present</returns>
	    private List<int> getZoomLevels()
        {
            JObject metadata = getMetadataObject();
            List<int> zoomLevels = new List<int>();
            if (metadata != null)
            {
                try
                {
                    string name = (string)metadata.GetValue(ConstantMbTilePlugin.KEY_NAME);
                    string version = (string)metadata.GetValue(ConstantMbTilePlugin.KEY_VERSION);
                    if (name != null && version != null)
                    {

                        string dirPath = this.mapDirectory + "\\" + version + "\\" + name;
                        using (IsolatedStorageFile isoFile = IsolatedStorageFile.GetUserStoreForApplication())
                        {
                            // create the file if not exists
                            if (isoFile.DirectoryExists(dirPath))
                            {
                                foreach (string filePath in isoFile.GetFileNames(dirPath))
                                {
                                    zoomLevels.Add(int.Parse(filePath));
                                }
                            }
                        }
                    }
                }
                catch (Exception)
                {
                    // Add log
                }
			   
            }   
		    return zoomLevels;
        }

        public override string executeStatement(String query, List<object> param)
        {
            string result = "";
            // not implemented
            return result;
        }

        public override directory_working_output getDirectoryWorking()
	    {
		    return new directory_working_output(getDirectory());
	    }

    }
}

File: src/android/MBTilesActionsFileImpl.java
package com.makina.offline.mbtiles;

import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.apache.cordova.CordovaResourceApi;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.net.Uri;
import android.os.Environment;
import android.util.Base64;
import android.util.Log;

/**
 * {@link IMBTilesActions} filesystem implementation.
 * 
 * @author <a href="mailto:sebastien.grimault@makina-corpus.com">S. Grimault</a>
 */
public class MBTilesActionsFileImpl extends MBTilesActionsGenImpl
{
	
	public MBTilesActionsFileImpl(Context context, CordovaResourceApi resourceApi, String typePath, String url) {
		super(context);
		
		if (typePath != null && typePath.equals(MBTilesPlugin.OPEN_TYPE_PATH_CDV)) {
			if (url == null || url.length() < 0) {
		 		url = "cdvfile://localhost/persistent/maps/";
			}
			
			Uri fileURL = resourceApi.remapUri(Uri.parse(url));
			mDirectory = fileURL.getPath();
			 
		} else if (typePath == null || typePath.equals(MBTilesPlugin.OPEN_TYPE_PATH_FULL)) {
			if (url == null || url.length() < 0) {
				url = Environment.getExternalStorageDirectory().getAbsolutePath() + "/Android/data/" +
						mContext.getPackageName() + "/maps/";
			}
			
			mDirectory = url;
		}
	}

	// declaration of path of file
	private File mapDirectory = null;
	
	@Override
	public void open(String name)
	{
		String path = getDirectory();
		Log.d(getClass().getName(), "try to open map directory '" + path + "'");
		this.mapDirectory = new File(path);
	}

	@Override
	public boolean isOpen()
	{
		return this.mapDirectory != null;
	}
	
	@Override
	public void close()
	{
		if (isOpen())
		{
			Log.d(getClass().getName(), "close '" + mapDirectory.getAbsolutePath() + "'");
			
			this.mapDirectory = null;
		}
	}

	@Override
	public JSONObject getMetadata()
	{
		File fileContentMedata = new File(this.mapDirectory.getAbsolutePath() + File.separator + "metadata.json");
		
		JSONObject metadata = new JSONObject();
		
		try
		{
			metadata = new JSONObject(FileUtils.readFileAsString(fileContentMedata));
		}
		catch (JSONException je)
		{
			Log.e(getClass().getName(), je.getMessage(), je);
		}
		catch (IOException ioe)
		{
			Log.e(getClass().getName(), ioe.getMessage(), ioe);
		}
		
		return metadata;
	}

	@Override
	public JSONObject getMinZoom()
	{
		JSONObject minZoom = new JSONObject();
		
		try
		{
			List<Integer> zoomLevels = getZoomLevels();
			
			if (!zoomLevels.isEmpty())
			{
				minZoom.put(KEY_MIN_ZOOM, zoomLevels.get(0));
			}
		}
		catch (JSONException je)
		{
			Log.e(getClass().getName(), je.getMessage(), je);
		}
		
		return minZoom;
	}

	@Override
	public JSONObject getMaxZoom()
	{
		JSONObject maxZoom = new JSONObject();
		
		try
		{
			List<Integer> zoomLevels = getZoomLevels();
			
			if (!zoomLevels.isEmpty())
			{
				maxZoom.put(KEY_MAX_ZOOM, zoomLevels.get(zoomLevels.size() - 1));
			}
		}
		catch (JSONException je)
		{
			Log.e(getClass().getName(), je.getMessage(), je);
		}
		
		return maxZoom;
	}

	@Override
	public JSONObject getTile(int zoomLevel, int column, int row)
	{
		JSONObject tileData = new JSONObject();
		JSONObject metadata = getMetadata();
		
		try
		{
			String name = (String) metadata.get(KEY_NAME);
			String version = (String) metadata.get(KEY_VERSION);
			String format = (String) metadata.get(KEY_FORMAT);
			
			// try to load the last zoom level if zoomLevel is too high
			int maxZoom = getMaxZoom().getInt(KEY_MAX_ZOOM);
			int currentZoomLevel = zoomLevel;
			
			if (zoomLevel > maxZoom)
			{
				currentZoomLevel = maxZoom;
			}
			
			File tileFile = new File(this.mapDirectory.getAbsolutePath() + File.separator + version + File.separator + name + File.separator + String.valueOf(currentZoomLevel) + File.separator + String.valueOf(column) + File.separator + String.valueOf(row) + "." + format);
			
			// test if file exist
			if (tileFile.exists())
			{
				ByteArrayOutputStream baos = new ByteArrayOutputStream();
				FileUtils.copyInputStream(new FileInputStream(tileFile), new BufferedOutputStream(baos));
				
				tileData.put(KEY_TILE_DATA, Base64.encodeToString(baos.toByteArray(), Base64.DEFAULT));

				baos.close();
			}
		}
		catch (JSONException je)
		{
			Log.e(getClass().getName(), je.getMessage(), je);
		}
		catch (FileNotFoundException fnfe)
		{
			Log.e(getClass().getName(), fnfe.getMessage(), fnfe);
		}
		catch (IOException ioe)
		{
			Log.e(getClass().getName(), ioe.getMessage(), ioe);
		}
		
		
		return tileData;
	}

	/**
	 * Tries to retrieve all available zoom levels for this MBTiles resource.
	 * @return a {@link List} of zoom levels as {@link Integer}
	 */
	private List<Integer> getZoomLevels()
	{
		JSONObject metadata = getMetadata();
		List<Integer> zoomLevels = new ArrayList<Integer>();
		
		try
		{
			String name = (String) metadata.get(KEY_NAME);
			String version = (String) metadata.get(KEY_VERSION);
			
			File contents = new File(this.mapDirectory.getAbsolutePath() + File.separator + version + File.separator + name);
			
			
			for (File zoomLevel : contents.listFiles())
			{
				zoomLevels.add(Integer.valueOf(zoomLevel.getName()));
			}
			
			Arrays.sort(zoomLevels.toArray(new Integer[]{}));
		}
		catch (JSONException je)
		{
			Log.e(getClass().getName(), je.getMessage(), je);
		}
		
		return zoomLevels;
	}
	
	@Override
	public JSONObject executeStatement(String query, String... params) {
		// not implemented
		JSONObject result = new JSONObject();
		return result;
	}

	/**
	 * return the directory of working action
	 * @param path : the pah of working
	 * @return <code>JSONObject</code>
	 */
	@Override
	public JSONObject getDirectoryWorking() {
		JSONObject directoryWorking = new JSONObject();
		try {
			directoryWorking.put(KEY_DIRECTORY_WORKING, getDirectory());
		} catch (JSONException e) {
			Log.w(getClass().getName(), e.getMessage());
		}
		return directoryWorking;
	}
	
}
